### ISSUE TITLE
[FEAT]: Deserialize strings to booleans

### ISSUE THREAD

**Issue by axos88** [GillianPerard/typescript-json-serializer#223]
### Description

When dealing with boolean properties, there should be a way to support deserializing a "true" or "false" string to the respective values

### Proposed solution

Maybe a standard beforeDeserialize hook in the library?

**Comment by axos88** [comment]
Probably something like:

```
function parseBoolean(input: any): boolean {
  if (input == "true") {
    return true;
  } else if (input == "false") {
    return false;
  } else {
    return input;
  }
}
```

**Comment by GillianPerard** [comment]
Hi, nice proposition, it could be a function that you pass to the `JsonProperty` that returns the `beforeDeserialize` and `afterSerialize` options.

```ts
@JsonObject()
class Test {
  @JsonProperty({ ...coerceBoolean(), name: 'my_bool' }) myBool: boolean;
}
```

Where the function could be something like

```ts
function coerceBoolean(options: { case: 'lower' | 'upper' | 'capitalize' } = { case: 'lower' }) {
  return {
    afterSerialize: (input: boolean) => string,
    beforeDeserialize: (input: string) => boolean
  }
}
```

**Comment by axos88** [comment]
I wouldnt want to necessarily serialize it into a string, only be able to accept inputs of different types. Think deserializing from url parameters, which are always strings.

**Comment by GillianPerard** [comment]
The problem is that some people wants to deserialize from an api and serialize to send back. So the types sent must be the same as the types received. Someone else would like to only deserialize the string into boolean but serialize the boolean (not into string) as in your case.

I have to provide a solution to satisfy everybody.

For example I can provide 2 functions (instead of one in my previous comment), to coerceBoolean (that you can use on `beforeDeserialize` and another one to transform back on `afterSerialize`.

**Comment by axos88** [comment]
Yeah, that's what I was suggesting. Instead of everyone rolling their own, either have a set of most commonly useful hooks in the library that can be used when applicable.

**Comment by GillianPerard** [comment]
What do you think about this:

```ts
export const tryCoerceBoolean = <T>(value: T): boolean | T => {
    if (isNullish(value) || isBoolean(value) || !isString(value)) {
        return value;
    }

    const stringValue = value.toString().toLocaleLowerCase();

    if (['true', '1'].includes(stringValue)) {
        return true;
    }

    if (['false', '0'].includes(stringValue)) {
        return false;
    }

    return value;
};
```
